forc-pkg/src/pkg.rs

```diff
@@ -193,6 +193,148 @@ impl BuildPlan {
+        let manifest_dep_pkgs = manifest
+            .deps()
+            .map(|(dep_name, dep)| {
+                // NOTE: Temporarily warn about `version` until we have support for registries.
+                if let Dependency::Detailed(det) = dep {
+                    if det.version.is_some() {
+                        println_yellow_err(&format!(
+                            "  WARNING! Dependency \"{}\" specifies the unused `version` field: \
+                        consider using `branch` or `tag` instead",
+                            dep_name
+                        ));
+                    }
+                }
+
+                let name = dep.package().unwrap_or(dep_name).to_string();
+                let source = dep_to_source(proj_path, dep)?;
+                let dep_pkg = Pkg { name, source };
+                Ok((dep_name, dep_pkg))
+            })
+            .collect::<Result<BTreeSet<_>>>()?;
+        let added: Vec<Pkg> = manifest_dep_pkgs
+            .difference(&plan_dep_pkgs)
+            .clone()
+            .map(|pkg| (pkg.1.clone()))
+            .collect();
+        let removed: Vec<Pkg> = plan_dep_pkgs
+            .difference(&manifest_dep_pkgs)
+            .clone()
+            .map(|pkg| (pkg.1.clone()))
+            .collect();
+
+        Ok((added, removed))
+    }
+
+    pub fn from_old_manifest(
+        &mut self,
+        manifest_file: &ManifestFile,
+        sway_git_tag: &str,
+        offline_mode: bool,
+    ) -> Result<Self> {
+        let mut graph = self.graph.clone();
+        let proj_node = *self
+            .compilation_order
+            .last()
+            .ok_or_else(|| anyhow!("Invalid Graph"))?;
+        let proj_id = graph[proj_node].id();
+        let (added, removed) =
+            self.get_difference_with_manifest(proj_node, manifest_file, sway_git_tag)?;
+
+        use petgraph::visit::{Bfs, Walker};
+        //find root_node_id
+        let root_node_id = *self.compilation_order().last().unwrap();
+        let visited_nodes: Vec<NodeIx> = Bfs::new(&graph, root_node_id).iter(&graph).collect();
+        for node in visited_nodes {
+            if node == root_node_id {
+                continue;
+            }
+            // if a -> b in the graph, a depends on b. if a is removed b will have 0
+            // incoming edge, if there is no other parent dependening on b remove it.
+            // to remove a either it will be in removed list or it needs to have 0 incoming edge.
+            if graph.edges_directed(node, Direction::Incoming).count() == 0
+                || removed
+                    .clone()
+                    .into_iter()
+                    .any(|removed_dep| removed_dep.name == graph[node].name)
+            {
+                graph.remove_node(node);
+            }
+        }
+
+        let mut visited_map: HashMap<Pinned, NodeIx> = HashMap::new();
+        // After removing some nodes the index of project node can change.
```

Comment:
Ahhh that's a great point - sorry I should have mentioned this quirk with the `petgraph::Graph` type earlier!

A much easier way of working around this (rather than recalculating the compilation order and reconstructing the path map, etc) is to use petgraph's `StableGraph` type rather than its regular `Graph` type.

The benefit of the `StableGraph` type over the `Graph` type is that you can remove nodes and edges without invalidating the indices of the remaining nodes and edges. When adding new nodes or edges, it will re-use free slots before extending the internal node or edge vectors. You can read a little more about `StableGraph` [here](https://docs.rs/petgraph/latest/petgraph/stable_graph/struct.StableGraph.html).

Knowing this, perhaps we should change our `Graph` type to:
```rust
pub type Graph = petgraph::stable_graph::StableGraph<Node, Edge, Directed, GraphIx>;
```
*This might require enabling a stable graph feature, I don't quite remember.*

Let me know once you've done this and simplified the following code as a result and I'll do another review :)